api/src/main/java/marquez/db/LineageDao.java

```diff
@@ -154,4 +171,51 @@ SELECT DISTINCT on(r.job_name, r.namespace_name) r.*, jv.version as job_version
       WHERE j.uuid in (<jobUuid>) OR j.symlink_target_uuid IN (<jobUuid>)
       ORDER BY r.job_name, r.namespace_name, created_at DESC""")
   List<Run> getCurrentRuns(@BindList Collection<UUID> jobUuid);
+
+  @SqlQuery(
+      """
+      WITH RECURSIVE
+        upstream_runs(
+                r_uuid, -- run uuid
+                dataset_uuid, dataset_version_uuid, dataset_namespace, dataset_name, -- input dataset version to the run
+                u_r_uuid, -- upstream run that produced that dataset version
+                depth -- current depth of traversal
+                ) AS (
+
+          -- initial case: find the inputs of the initial runs
+          select r.uuid,
+                 dv.dataset_uuid, dv."version", dv.namespace_name, dv.dataset_name,
+                 dv.run_uuid,
+                 0 AS depth -- starts at 0
+          FROM (SELECT :runId::uuid AS uuid) r -- initial run
+          LEFT JOIN runs_input_mapping rim ON rim.run_uuid = r.uuid
+          LEFT JOIN dataset_versions dv ON dv.uuid = rim.dataset_version_uuid
+
+        UNION
+
+          -- recursion: find the inputs of the inputs found on the previous iteration and increase depth to know when to stop
+          SELECT
+                ur.u_r_uuid,
+                dv2.dataset_uuid, dv2."version", dv2.namespace_name, dv2.dataset_name,
+                dv2.run_uuid,
+                ur.depth + 1 AS depth -- increase depth to check end condition
+          FROM upstream_runs ur
+          LEFT JOIN runs_input_mapping rim2 ON rim2.run_uuid = ur.u_r_uuid
+          LEFT JOIN dataset_versions dv2 ON dv2.uuid = rim2.dataset_version_uuid
+          -- end condition of the recursion: no input or depth is over the maximum set
+          -- also avoid following cycles (ex: merge statement)
+          WHERE ur.u_r_uuid IS NOT NULL AND ur.u_r_uuid <> ur.r_uuid AND depth < :depth
```

Comment:
Have you considered any anti-join step within recursion step to avoid following cycles? Something like:
```
LEFT JOIN upstream_runs eur ON ur.u_r_uuid = eur.r_uuid WHERE eur.r_uuid is NULL
```

Otherwise we may expect recursive query result to explode. Although we do `distinct` at the bottom, this may still affect performance.

Reply:
I see your point. That would avoid following the same edge twice even when there's no cycle. Today it happens when there is a diamond in the graph for example. This is why I added a distinct statement at the end.
On the other end it does add an extra join on the temporary recursive table, so it's probably a trade-off in term of performance. It might slow down in the common case to speed up some corner cases. Although the temporary table upstream_runs is small so not by much.

I think this might be unnecessary since I think the only cycle that can happen on runs are pointing to themselves (this happens with the SQL merge statement when a table is an input and an output of the job). 
Here we avoid following them with "ur.u_r_uuid <> ur.r_uuid" 

I think I can add the statement you are suggesting. Then I would remove the DISTINCT constraint in the end as we'd prevent completely following the same edge twice.

What do you think?

Reply:
Yeah, you're right it's just a corner case when upstream nodes are visited. 
This becomes a problem when one traverses both upstream and downstream edges at the same time and can reach the same node with different depth oftentimes.